opencog/atoms/core/ImplicationLink.cc

```diff
@@ -61,26 +61,23 @@ ImplicationLink::ImplicationLink(Link &l)
 void ImplicationLink::extract_variables(const HandleSeq& oset)
 {
 	size_t sz = _outgoing.size();
-	// For now we assume no implicit scoping, so the number of
-	// arguments is 2 when sugar syntax isn't used (no scoping), or 3
-	// when the sugar syntax (scoping) is used.
+
 	if (sz < 2 or 3 < sz)
 		throw InvalidParamException(TRACE_INFO,
-			"Expecting an outgoing set size of at most two, got %d", sz);
+			"Expecting an outgoing set of size 2 or 3, got %d", sz);
+
+	// If there is an explicit variable declaration, then use that.
+	// If there is no variable declaration, then assume that all of
+	// the free variables in the link are scoped.
+	ScopeLink::extract_variables(oset);
```

Comment:
If I remove ImplicationLink.cc and ImplicationLink.h entirely, then only two unit tests fail: 
```
	 18 - MapLinkUTest (SEGFAULT)
	 92 - BackwardChainerUTest (Failed)
```

The MapLinkUTest fails because .. well, let me explain below ... but the shorter question is: do we need ImplicationLink.cc at all? Can we just get rid of it entirely?  If we did, then I think it would solve the problem that you are having.